flows/analysis/build_defs.bzl

```diff
@@ -0,0 +1,120 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Design analysis utilities written using modular hardware flows."""
+
+load("//flows:flows.bzl", "flow_binary", "tcl_script_prefix")
+load("//flows/openroad:build_defs.bzl", "assemble_openroad_step", "read_standard_cells")
+load("//flows/yosys:build_defs.bzl", "yosys_synth_file_step")
+load("//pdk:build_defs.bzl", "StandardCellInfo")
+
+def _analyze_netlist_step_impl(ctx):
+    tcl_commands = [tcl_script_prefix]
+
+    (read_cells_commands, read_cells_runfiles) = read_standard_cells(ctx)
+
+    tcl_commands.extend(read_cells_commands)
+
+    base_script = ctx.actions.declare_file(ctx.attr.name + "_base.tcl")
+    full_script = ctx.actions.declare_file(ctx.attr.name + ".tcl")
+
+    ctx.actions.write(
+        output = base_script,
+        content = "\n".join(tcl_commands + ["\n"]),
+    )
+
+    ctx.actions.run_shell(
+        inputs = [base_script, ctx.file.analysis_script],
+        outputs = [full_script],
+        command = "cat {base_script} {analysis_script} > {full_script}".format(
+            base_script = base_script.path,
+            analysis_script = ctx.file.analysis_script.path,
+            full_script = full_script.path,
+        ),
+    )
+
+    return assemble_openroad_step(
+        ctx,
+        ctx.attr.name,
+        full_script,
+        read_cells_runfiles,
+        inputs = ["netlist"] + ctx.attr.inputs,
+        outputs = ctx.attr.outputs,
+        constants = ["top"] + ctx.attr.outputs,
+    )
+
+analyze_netlist_step = rule(
+    implementation = _analyze_netlist_step_impl,
+    attrs = {
+        "_openroad": attr.label(
+            default = Label("@org_theopenroadproject//:openroad"),
+            executable = True,
+            cfg = "exec",
+        ),
+        "analysis_script": attr.label(
+            doc = "OpenROAD Tcl script implementing the analysis",
+            allow_single_file = [".tcl"],
+            mandatory = True,
+        ),
+        "standard_cells": attr.label(
+            doc = "Standard cells to use for analysis",
+            providers = [StandardCellInfo],
+            mandatory = True,
+        ),
+        "inputs": attr.string_list(
+            doc = "Additional file inputs (beyond netlist) required by this analysis step",
+        ),
+        "outputs": attr.string_list(
```

Comment:
OOC, why string_list and not output_list?

Reply:
These are just handles to the inputs, outputs and constants, not the files and string constants themselves. Those will be hooked up in two different ways:
1. Inside bazel by the run_flow rule when it is run on a sequence of flow steps.
2. Outside bazel by the wrapper scripts generated by the flow_binary rule (and the underlying environment variables provided by the user).

I've updated the documentation to make that a bit clearer.